### PR TITLE
(QTM-189) - Events Keypress DropdownLight

### PR DESCRIPTION
**Pull request by douglasjr97**
## Description
https://jirasoftware.catho.com.br/browse/QTM-189

## Review guide
- [ ] Unit tests (yarn test:components)
- [ ] Code review
- [ ] Documentation

### CODE REVIEW

**Review by MarcosViniciusPC — CHANGES_REQUESTED** [review]
[] code review
[] unit

When I press the "Enter" button when the button is focused the list of items is not opened, please adjust this.
Also look at all the points mentioned below

File: components/DropdownLight/SubComponents/UseKeyPress.jsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+  useEffect(() => {
+    window.addEventListener('keydown', downHandler);
+    window.addEventListener('keyup', upHandler);
+
+    return () => {
+      window.removeEventListener('keydown', downHandler);
+      window.removeEventListener('keyup', upHandler);
+    };
+  });
```
[the remark this answers is not in the document]
Comment (douglasjr97) [reply components/DropdownLight/SubComponents/UseKeyPress.jsx:28]: Agree, when we don't use the empty array, the component renders multiple times

File: components/DropdownLight/DropdownLight.unit.test.jsx [hunk 2]
```diff
@@ -81,4 +99,61 @@ describe('<DropdownLight />', () => {
...
+
+    expect(onChangeMock).toHaveBeenCalledTimes(1);
+    expect(onChangeMock).toHaveBeenCalledWith('Banana');
+  });
+
+  it('should open select when user press enter on it', () => {
+    render(<DropdownLight items={itemsStringMock} />);
+
+    userEvent.tab();
+    userEvent.keyboard('{Enter}');
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.unit.test.jsx:144]: Could you put the "{Enter}" in a global variable, what do you think?

File: components/DropdownLight/DropdownLight.unit.test.jsx [hunk 3]
```diff
@@ -81,4 +99,61 @@ describe('<DropdownLight />', () => {
...
+    ).toBeInTheDocument();
+
+    userEvent.keyboard('{Enter}');
+
+    expect(
+      screen.getByRole('button', { name: 'abrir lista de itens' }),
+    ).toBeInTheDocument();
+    expect(
+      screen.queryByRole('button', { name: 'fechar lista de itens' }),
+    ).not.toBeInTheDocument();
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.unit.test.jsx:157]:
```suggestion
```
I think this code snippet is unnecessary.

File: components/DropdownLight/DropdownLight.unit.test.jsx [hunk 4]
```diff
@@ -61,9 +73,15 @@ describe('<DropdownLight />', () => {
...
-    const dropdown = screen.getByRole('button');
+    const dropdown = screen.getByRole('button', {
+      name: 'abrir lista de itens',
+    });
     userEvent.click(dropdown);
 
+    expect(
+      screen.getByRole('button', { name: 'fechar lista de itens' }),
+    ).toBeInTheDocument();
+
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.unit.test.jsx:84]:
```suggestion
```

File: components/DropdownLight/DropdownLight.unit.test.jsx [hunk 5]
```diff
@@ -61,9 +73,15 @@ describe('<DropdownLight />', () => {
   it('should show the check icon when the item from list selected', () => {
     render(<DropdownLight items={itemsStringMock} />);
 
-    const dropdown = screen.getByRole('button');
+    const dropdown = screen.getByRole('button', {
+      name: 'abrir lista de itens',
+    });
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.unit.test.jsx:78]: nice!

File: components/DropdownLight/DropdownLight.unit.test.jsx [hunk 6]
```diff
@@ -44,9 +50,15 @@ describe('<DropdownLight />', () => {
...
     userEvent.click(dropdown);
 
+    expect(
+      screen.getByRole('button', { name: 'fechar lista de itens' }),
+    ).toBeInTheDocument();
+
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.unit.test.jsx:61]:
```suggestion
```

File: components/DropdownLight/DropdownLight.unit.test.jsx [hunk 7]
```diff
@@ -27,9 +27,15 @@ describe('<DropdownLight />', () => {
...
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.unit.test.jsx:38]:
```suggestion
```

Comment (MarcosViniciusPC) [reply components/DropdownLight/DropdownLight.unit.test.jsx:38]: I think it's outside the scope of the test

File: components/DropdownLight/DropdownLight.unit.test.jsx (hunk 6 above)
Comment (MarcosViniciusPC) [reply components/DropdownLight/DropdownLight.unit.test.jsx:61]: I think it's outside the scope of the test

File: components/DropdownLight/DropdownLight.unit.test.jsx (hunk 4 above)
Comment (MarcosViniciusPC) [reply components/DropdownLight/DropdownLight.unit.test.jsx:84]: I think it's outside the scope of the test

File: components/DropdownLight/SubComponents/UseKeyPress.jsx [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    return () => {
+      window.removeEventListener('keydown', downHandler);
+      window.removeEventListener('keyup', upHandler);
+    };
+  }, []);
+
+  return keyPressed;
+};
+
+export default useKeyPress;
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/SubComponents/UseKeyPress.jsx:33]: I think it's interesting to make a spike to think about ways to improve this hook's performance. Apparently every time the external component that uses it is rendered it creates a new listener for keyboard events and I believe this is a problem. Perhaps the solution is to implement a useCallback in the parent component.

**Review by MarcosViniciusPC — CHANGES_REQUESTED** [review]
[] Code review
[] Unit

When selecting an item from the list, the focus must go back to the dropdown (button), according to the behavior of the old dropdown;

**Review by MarcosViniciusPC — CHANGES_REQUESTED** [review]
[] Code review
[] Unit

File: components/DropdownLight/DropdownLight.jsx [hunk 9]
```diff
@@ -141,62 +142,153 @@ const SelectionListItem = styled.li`
...
 `;
 
-const DropdownLight = ({ disabled, items, theme, placeholder, name }) => {
+const DropdownLight = ({
+  disabled,
+  items,
+  theme,
+  placeholder,
+  name,
+  onChange,
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.jsx:151]:
```suggestion
```

File: components/DropdownLight/DropdownLight.jsx [hunk 10]
```diff
@@ -141,62 +142,153 @@ const SelectionListItem = styled.li`
...
+
+    if (isOpen) {
+      listOptions.current.children[0].focus();
+    }
+  };
+
+  const selectItem = item => {
     setSelectedItem(item?.value || item);
     setItemLabel(item?.label || item);
+    onChange(item?.value || item);
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.jsx:179]:
```suggestion
```

File: components/DropdownLight/DropdownLight.jsx [hunk 11]
```diff
@@ -210,6 +302,7 @@ DropdownLight.propTypes = {
     baseFontSize: PropTypes.number,
   }),
   name: PropTypes.string,
+  onChange: PropTypes.func,
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.jsx:305]:
```suggestion
```

File: components/DropdownLight/DropdownLight.jsx [hunk 12]
```diff
@@ -221,6 +314,7 @@ DropdownLight.defaultProps = {
   },
   placeholder: 'Select an option',
   name: '',
+  onChange: () => {},
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.jsx:317]:
```suggestion
```

File: components/DropdownLight/DropdownLight.unit.test.jsx [hunk 13]
```diff
@@ -81,4 +85,37 @@ describe('<DropdownLight />', () => {
...
+  it('should close Dropdown Options when user press Escape', () => {
+    render(<DropdownLight items={itemsStringMock} />);
+
+    const dropdown = screen.getByRole('button');
+    userEvent.click(dropdown);
+
+    userEvent.keyboard(`${EscapeKeyCode}`);
+
+    expect(screen.queryAllByRole('option')).toHaveLength(0);
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.unit.test.jsx:97]:
```suggestion
    expect(screen.getByRole('list')).not.toBeInTheDocument();
```

File: components/DropdownLight/DropdownLight.unit.test.jsx [hunk 14]
```diff
@@ -81,4 +85,37 @@ describe('<DropdownLight />', () => {
...
+
+    expect(
+      screen.getByRole('button', { name: 'abrir lista de itens' }),
+    ).toBeInTheDocument();
+  });
+
+  it('using only keyboard', () => {
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.unit.test.jsx:104]:
```suggestion
  it('should allow User to select an option using only keyboard', () => {
```

File: components/DropdownLight/DropdownLight.unit.test.jsx [hunk 15]
```diff
@@ -81,4 +85,37 @@ describe('<DropdownLight />', () => {
...
+    const bananaItem = itemsStringMock[1];
+
+    const input = screen.getByRole('textbox', { hidden: true });
+
+    userEvent.keyboard(
+      `${ArrowDownKeyCode}${ArrowDownKeyCode}${ArrowUpKeyCode}${EnterKeyCode}`,
+    );
+
+    expect(input.value).toEqual(bananaItem);
+    expect(screen.queryByRole('option')).not.toBeInTheDocument();
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.unit.test.jsx:119]:
```suggestion
    expect(screen.getByRole('list')).not.toBeInTheDocument();
```

File: components/DropdownLight/DropdownLight.unit.test.jsx [hunk 16]
```diff
@@ -81,4 +85,37 @@ describe('<DropdownLight />', () => {
...
+    expect(
+      screen.getByRole('button', { name: 'abrir lista de itens' }),
+    ).toBeInTheDocument();
+  });
+
+  it('using only keyboard', () => {
+    render(<DropdownLight items={itemsStringMock} />);
+
+    userEvent.tab();
+    userEvent.keyboard(`${EnterKeyCode}`);
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.unit.test.jsx:108]:
```suggestion
    userEvent.keyboard(`${EnterKeyCode}`);
    
    expect(screen.getByRole('list')).toBeInTheDocument();
```

File: components/DropdownLight/DropdownLight.unit.test.jsx [hunk 17]
```diff
@@ -81,4 +85,37 @@ describe('<DropdownLight />', () => {
...
+
+    const bananaItem = itemsStringMock[1];
+
+    const input = screen.getByRole('textbox', { hidden: true });
+
+    userEvent.keyboard(
+      `${ArrowDownKeyCode}${ArrowDownKeyCode}${ArrowUpKeyCode}${EnterKeyCode}`,
+    );
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.unit.test.jsx:116]:
```suggestion
    userEvent.keyboard(
      `${ArrowDownKeyCode}${ArrowDownKeyCode}${ArrowUpKeyCode}${EnterKeyCode}`,
    );
    const bananaItem = itemsStringMock[1];
    const input = screen.getByRole('textbox', { hidden: true });
```

**Review by MarcosViniciusPC — CHANGES_REQUESTED** [review]
[] Code review
[x] Unit

File: components/DropdownLight/DropdownLight.unit.test.jsx (hunk 16 above)
Comment (douglasjr97) [reply components/DropdownLight/DropdownLight.unit.test.jsx:108]: Why should I return a list if I'm simulating a keyboard selection?

File: components/DropdownLight/DropdownLight.unit.test.jsx (hunk 17 above)
Comment (douglasjr97) [reply components/DropdownLight/DropdownLight.unit.test.jsx:116]: Why should I return a list if I'm simulating a keyboard selection?

**Review by MarcosViniciusPC — APPROVED** [review]
[x] Code review
[x] Unit
[x] Documentation

**Review by rapahaeru — APPROVED** [review]

File: components/DropdownLight/DropdownLight.unit.test.jsx [hunk 18]
```diff
@@ -81,4 +85,39 @@ describe('<DropdownLight />', () => {
...
+
+    const dropdown = screen.getByRole('button');
+    userEvent.click(dropdown);
+
+    userEvent.keyboard(EscapeKeyCode);
+
+    expect(screen.queryByRole('list')).not.toBeInTheDocument();
+
+    expect(
+      screen.getByRole('button', { name: 'abrir lista de itens' }),
```
[the remark this answers is not in the document]
Comment (douglasjr97) [reply components/DropdownLight/DropdownLight.unit.test.jsx:100]: The open or closed "area-label" is being used on the button, this verification is only to know if the list is not on the screen

**Pull request merged**